### ISSUE TITLE
replica-side load shedding is not aggressive enough

### ISSUE THREAD

**Issue by denesb** [scylladb/scylladb#14909]
We have observed, in multiple issues that when a shard on a replica is overloaded, it breaks down completely, throughput collapses, latencies go through the roof and the node/shard can even become completely unresponsive to new connection attempts.
Example issue: https://github.com/scylladb/scylladb/issues/13759.

I believe the root cause is that we have almost nonexistent load shedding on the replica side. When reads arrive, they have to wait for admission on the reader concurrency semaphore. If the node is overloaded, the reads will be queued. They can time out while in the queue, but will not time out once admitted. Once the shard is sufficiently loaded, it is possible that most queued reads will time out, because the average time it takes to for a queued read to be admitted is around that of the timeout. Reads admitted will possibly also only finish after their timeout expires.

We have coordinator side load-shedding, but when only a single replica is pushed over the limit, this is not enough. With speculative reads, the overloaded replica will be worked around, the other nodes shouldering the load. But the overloaded replica will be continued to be picked for new reads and therefore it will be hammered.

I think we need to do the following, to address this:
* keep mean time-to-finish statistic in semaphore
* when qeueing a read: if `mean_time * queue_size >= timeout * X` (I propose 0.5 for `X`), reject the read on the spot
* if a read times out in the queue: check the above condition again (`mean_time` could have changed since the queue was last checked), if it fails, kill reads from the front of the queue, such that the invariant is maintained
* add timeout check for all ongoing reads, maybe all readers should check timeout on each `fill_buffer()` call

**Comment by denesb** [comment]
Also, we should probably not use exceptions for timeout. Maybe push down the result system used int he coordiantor code, all the way to the semaphore. Not sure how viable is this, but exceptions do pile on an overloaded shard, at the worst time.

**Comment by denesb** [comment]
Alternative solution: make the RPC layer aware of the semaphore queue size. When above a certain limit, stop reading from the network (for a certain list of verbs). The semaphore could then signal to the RPC layer that new reads can be admitted, when the situation improves. Some amount of queued reads *are* desired, so we should not aim for an empty queue.
The RPC layer is already timeout aware and dropping already timed out reads as they are read from the network is much cheaper then the semaphore dropping it later.

**Comment by denesb** [comment]
> What will prevent the coordinator from continuing to hammer that poor replica?

We should maybe make changes on the coordinator side too. But we cannot rely on a well-behaved coordinator. I think we should make the replica be able to continue serving at (or close to) max throughput, even when overloaded.

**Comment by mykaul** [comment]
BTW, I wonder if random early drop (perhaps weighted? according to service level or some other criteria) wouldn't be better than the above. It might also prevent multiple busy replicas throwing out the same reads?

**Comment by denesb** [comment]
> BTW, I wonder if random early drop (perhaps weighted? according to service level or some other criteria) wouldn't be better than the above. It might also prevent multiple busy replicas throwing out the same reads?

Different service levels are on different connections and different semaphores. Multiple busy replicas throwing out the same read is actually desirable. A single replica throwing out the read, fails the read altogether, so the work done by the remaining replicas is thrown away.

**Comment by avikivity** [comment]
What's the ratio of queued reads to executing reads?

If the ratio is high, then we can address the issue on the queue side (easy).

If it's low, then it means we're admitting too many reads. Once we admit a read getting rid of it is expensive, since it already did some work.

Assuming the ratio is high then we can look at the head of the queue. If the read waiting there is already halfway towards its timeout, then this is similar to your second condition. It's made somewhat more complicated by the fact that different reads can have different timeouts.

One thing we can do is to reverse the processing from FIFO to LIFO (@tgrabiec proposed it a long time ago). A FIFO read already had a large part of its timeout wasted, so we may as well give up on it and start a fresh read that has higher probablity of completing.

A problem common to all replica-side load shedding is that if two replicas shed different requests, then we end up increasing the damage. Let's say replica A sheds a random 50% and replica B sheds a random 50%, only 25% of requests to A+B will have a response. So we need the coordinator to influence the shedding (perhaps by selecting a random number and attaching it to the request; replicas will prefer to shed requests with higher numbers and so coordinate the shedding). @piodul.

**Comment by denesb** [comment]
The ratio is very high, when a shard is overloaded, there is usually thousands of queued reads, while the executing reads are 10 or less (usually 1-5).

**Comment by tgrabiec** [comment]
Refs https://github.com/scylladb/scylladb/issues/6428

**Comment by mykaul** [comment]
Do we have a single queue, for reads with different timeouts?

**Comment by denesb** [comment]
> Do we have a single queue, for reads with different timeouts?

Yes, all reads that have the same scheduling group, share the same queue, in all the different stages of the read pipeline.

**Comment by mykaul** [comment]
Moving to 6.1, but setting release blocker, so it'll get more attention.

**Comment by avikivity** [comment]
> Moving to 6.1, but setting release blocker, so it'll get more attention.

Did it work?

**Comment by mykaul** [comment]
> > Moving to 6.1, but setting release blocker, so it'll get more attention.
> 
> Did it work?

I've always looked at it and what else is on the table. From that perspective, yes, it was not forgotten.
Was it prioritised? No.
